Sources/NIOCore/AsyncChannelAdapter.swift

```diff
@@ -0,0 +1,514 @@
+            case .notClosed:
+                self = .writeClosed
+                return .nothing
+            case .readClosed:
+                self = .bothClosed
+                return .close
+            case .writeClosed, .bothClosed:
+                preconditionFailure("Duplicate write closure")
+            }
+        }
+    }
+
+    @usableFromInline
+    enum Action {
+        case nothing
+        case close
+    }
+
+    @usableFromInline
+    var _state: State
+
+    @inlinable
+    init() {
+        self._state = .notClosed
+    }
+
+    @inlinable
+    func closeRead() -> Action {
+        return self._state.closeRead()
+    }
+
+    @inlinable
+    func closeWrite() -> Action {
+        return self._state.closeWrite()
+    }
+}
+
+/// A `ChannelHandler` that is used to transform the inbound portion of a NIO
+/// `Channel` into an `AsyncSequence` that supports backpressure.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+@usableFromInline
+internal final class NIOAsyncChannelAdapterHandler<InboundIn: Sendable>: ChannelDuplexHandler {
+    @usableFromInline
+    typealias OutboundIn = Any
+
+    @usableFromInline
+    typealias OutboundOut = Any
+
+    @usableFromInline
+    typealias Source = NIOThrowingAsyncSequenceProducer<InboundIn, Error, NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark, NIOAsyncChannelAdapterHandler<InboundIn>.Delegate>.Source
+
+    @usableFromInline var source: Source?
+
+    @usableFromInline var context: ChannelHandlerContext?
+
+    @usableFromInline var buffer: [InboundIn] = []
+
+    @usableFromInline var producingState: ProducingState = .keepProducing
+
+    @usableFromInline let loop: EventLoop
+
+    @usableFromInline let closeRatchet: CloseRatchet
+
+    @inlinable
+    init(loop: EventLoop, closeRatchet: CloseRatchet) {
+        self.loop = loop
+        self.closeRatchet = closeRatchet
+    }
+
+    @inlinable
+    func handlerAdded(context: ChannelHandlerContext) {
+        self.context = context
+    }
+
+    @inlinable
+    func handlerRemoved(context: ChannelHandlerContext) {
+        self._completeStream(context: context)
+        self.context = nil
+    }
+
+    @inlinable
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        self.buffer.append(self.unwrapInboundIn(data))
+
+        // We forward on reads here to enable better channel composition.
+        context.fireChannelRead(data)
+    }
+
+    @inlinable
+    func channelReadComplete(context: ChannelHandlerContext) {
+        self._deliverReads(context: context)
+        context.fireChannelReadComplete()
+    }
+
+    @inlinable
+    func channelInactive(context: ChannelHandlerContext) {
+        self._completeStream(context: context)
+        context.fireChannelInactive()
+    }
+
+    @inlinable
+    func errorCaught(context: ChannelHandlerContext, error: Error) {
+        self._completeStream(with: error, context: context)
+        context.fireErrorCaught(error)
```

Comment:
Should we have an option here to not propagate the error further down the pipeline so the code handling the `NIOAsyncChannel.inboundStream` can deal with the error instead. Currently it is propagated to the next channel handler `NIOAsyncChannelWriterHandler` which closes the write side of the channel.

Reply:
The writer handler should probably just not choke on that error.

Reply:
That's an option as well